File: src/main/java/bernard/tatin/Tools/ThMemoryFiller.java
package bernard.tatin.Tools;

import bernard.tatin.Constants.ApplicationConstants;
import bernard.tatin.Threads.Mutex;
import bernard.tatin.Threads.ThConsumer;
import bernard.tatin.Threads.ThPrinterClient;

import java.util.Arrays;
import java.util.stream.Stream;

public class ThMemoryFiller extends ThPrinterClient implements ThConsumer, Runnable {
    public final static ThMemoryFiller mainMemoryFiller = new ThMemoryFiller();
    private final Mutex mutex = new Mutex();
    private Byte[] memory = null;
    private long memory_size = 0;

    private ThMemoryFiller() {
    }

    public void consume() {
        while (true) {
            try {
                memory = memory != null ?
                        Stream.concat(Arrays.stream(memory),
                                fillMemory(ApplicationConstants.MEMORY_INCREMENT)).
                                toArray(Byte[]::new) :
                        fillMemory(ApplicationConstants.MEMORY_INCREMENT).
                                toArray(Byte[]::new);

            } catch (OutOfMemoryError e) {
                printError("ERROR (ThMemoryFiller::consume): " + e.toString());
                memory = null;
            }

            setMemorySize();
            try {
                Thread.sleep(100L);
            } catch (InterruptedException e) {
                printError("ERROR InterruptedException : " + e.getMessage());
            }
        }
    }

    public long getMemorySize() {
        long rmem = 0;
        try {
            mutex.lock();
        } catch (Exception e) {
            printError("ERROR : cannot acquire mutex " + e.toString());
        }
        rmem = memory_size;
        mutex.unlock();
        return rmem;
    }

    private void setMemorySize() {
        try {
            mutex.lock();
        } catch (InterruptedException e) {
            printError("ERROR : cannot acquire mutex " + e.toString());
        }
        memory_size = 0;
        if (memory != null) {
            memory_size = memory.length;
        }
        mutex.unlock();
    }

    public void initialize() {
        Thread theMemoryFiller = new Thread(this, "ThMemoryFiller");

        theMemoryFiller.setDaemon(true);
        theMemoryFiller.start();
    }

    public synchronized void run() {
        consume();
    }

    private Stream<Byte> fillMemory(int bytes) {
        return Stream.iterate((byte) 0,
                b -> (byte) ((b + (byte) 1) % (byte) 253)).limit(bytes);
    }
}

File: kotlin-version/src/main/kotlin/bernard/tatin/Tools/ThMemoryFiller.kt
package bernard.tatin.Tools

import bernard.tatin.Constants.ApplicationConstants
import bernard.tatin.Threads.*

import java.util.Arrays
import java.util.stream.Stream

class ThMemoryFiller private constructor() : AThConsumer(), IThPrinterClient {
    private val mutex = Mutex()
    private var memory: ByteArray? = null
    private var theMemorySize: Long = 0
    val memory_unit = ByteArray(ApplicationConstants.MEMORY_INCREMENT,
            { b -> ((b + 1.toByte()) % 253.toByte()).toByte() })

    fun memorySize(): Long {
            val rmem: Long
            try {
                mutex.lock()
            } catch (e: Exception) {
                printError("ERROR : cannot acquire mutex " + e.toString())
            }

            rmem = theMemorySize
            mutex.unlock()
            return rmem
        }
    private fun setMemorySize() {
        try {
            mutex.lock()
        } catch (e: InterruptedException) {
            printError("ERROR : cannot acquire mutex " + e.toString())
        }

        theMemorySize = 0
        if (memory != null) {
            theMemorySize = memory!!.size.toLong()
        }
        mutex.unlock()
    }

    fun concatArrays() : ByteArray {
        val ab = ByteArray(theMemorySize.toInt() + ApplicationConstants.MEMORY_INCREMENT)
        System.arraycopy(memory, 0, ab, 0, theMemorySize.toInt())
        System.arraycopy(memory_unit, 0, ab, theMemorySize.toInt(), ApplicationConstants.MEMORY_INCREMENT)
        return ab
    }
    override fun innerLoop() {
        try {
            memory = if (memory != null) {
                concatArrays()
            } else {
                memory_unit
            }
        } catch (e: OutOfMemoryError) {
            printError("ERROR (ThMemoryFiller::consume): " + e.toString())
            memory = null
        }

        setMemorySize()
        try {
            Thread.sleep(100L)
        } catch (e: InterruptedException) {
            printError("ERROR InterruptedException : " + e.toString())
        }

    }


    override fun printString(str: String) {
        ThPrinter.mainInstance.printString(str)
    }

    override fun printStrings(strings: Array<String>) {
        ThPrinter.mainInstance.printStrings(strings)
    }

    override fun printError(str: String) {
        ThPrinter.mainInstance.printError(str)
    }

    companion object {
        val mainInstance = ThMemoryFiller()
    }
}
